src/util/submit-addon.js

```diff
@@ -0,0 +1,388 @@
+    return this.getJson(response);
+  }
+
+  doNewAddonOrVersionSubmit(
+    addonId: string,
+    uuid: string,
+    metaDataJson: Object,
+  ): Promise<typeof Response> {
+    const url = `${this.apiUrlPrefix}addon/${addonId}/`;
+    const jsonData = { version: { upload: uuid }, ...metaDataJson };
+    return this.fetch(url, 'PUT', JSON.stringify(jsonData));
+  }
+
+  waitForApproval(addonId: string, versionId: number): Promise<string> {
+    let approvalCheckTimeout;
+    const detailUrl =
+      `${this.apiUrlPrefix}addon/${addonId}/versions/${versionId}/`;
+
+    return new Promise((resolve, reject) => {
+      const abortTimeout = setTimeout(() => {
+        clearTimeout(approvalCheckTimeout);
+
+        reject(new Error('Approval Timeout.'));
+      }, this.approvalCheckTimeout);
+
+      const pollApprovalStatus = async () => {
+        try {
+          const detailResponse = await this.fetch(detailUrl);
+          if (!detailResponse.ok) {
+            return reject(new Error('Getting addon details failed.'));
+          }
+          const detailResponseData = await detailResponse.json();
+
+          const file = detailResponseData.file;
+          if (file && file.status === 'public') {
+            clearTimeout(abortTimeout);
+            resolve(file.url);
+          } else {
+            // The add-on hasn't been approved yet, so wait for a while and try again.
+            approvalCheckTimeout = setTimeout(
+              pollApprovalStatus,
+              this.approvalCheckInterval
+            );
+          }
+        } catch (err) {
+          clearTimeout(abortTimeout);
+          reject(err);
+        }
+      };
+
+      pollApprovalStatus();
+    });
+  }
+
+  async getJson(response: typeof Response): Promise<any> {
+    if (response.status < 100 || response.status >= 500) {
+      return new Promise((resolve, reject) => {
+        reject(new Error(`Getting response failed: ${response.status}.`));
+      });
+    } else {
+      const data = await response.json();
+      return new Promise((resolve, reject) => {
+        if (!response.ok) {
+          this.logger.log(data);
+          reject(new Error('Bad Request.'));
+        } else {
+          resolve(data);
+        }
+      });
+    }
+  }
+
+  fetch(
+    url: string,
+    method: string = 'GET',
+    body?: typeof FormData | string,
+    jwt: typeof defaultJwt = defaultJwt
+  ): Promise<typeof Response> {
+    const authToken = jwt.sign({ iss: this.apiKey }, this.apiSecret, {
+      algorithm: 'HS256',
+      expiresIn: this.apiJwtExpiresIn,
+    });
+
+    this.logger.log(`Fetching URL: ${url}`);
+    let headers;
+    if (typeof body === 'string') {
+      headers = {
+        'Authorization': `JWT ${authToken}`,
+        'Accept': 'application/json',
+        'Content-Type': 'application/json',
+      };
+    } else {
+      headers = {
+        Authorization: `JWT ${authToken}`,
+        Accept: 'application/json',
+      };
+    }
```

Comment:
Using `let headers` raises an eslint error: "'headers' is never reassigned. Use 'const' instead."
but using `const header` raises a flow error "Cannot assign 'application/json' to headers['Content-Type'] because property Content-Type is missing in object literal"

Reply:
Now I remember why I didn't like Flow...

New suggestion (pseudo-code):

```js
let headers = {...};
if (typeof body === 'string') {
  headers = {
    ..headers,
    'content-type': 'application/json',
  };
}
```

Reply:
👍 works!